prov/efa/src/efa_ep.c

```diff
@@ -611,6 +633,132 @@ static struct fi_ops efa_ep_ops = {
 	.ops_open = fi_no_ops_open,
 };
 
+static void efa_ep_progress_internal(struct efa_cq *efa_cq, uint64_t flags)
+{
+	struct util_cq *cq = &efa_cq->util_cq;
+	int cq_entries = 500;
+	int i;
+	ssize_t ret;
+	uint8_t cq_entry[(efa_cq->entry_size / sizeof(uint8_t)) * cq_entries];
+	struct fi_cq_tagged_entry *temp_cq_entry;
+	struct fi_cq_err_entry cq_err_entry;
+	fi_addr_t src_addr[cq_entries];
+
+	VALGRIND_MAKE_MEM_DEFINED(&cq_entry, sizeof(cq_entry));
+
+	ret = efa_cq_readfrom(&cq->cq_fid, cq_entry, cq_entries,
+			      (flags & FI_SOURCE) ? src_addr : NULL);
+	if (ret == -FI_EAGAIN)
+		return;
+
+	if (OFI_UNLIKELY(ret < 0)) {
+		ret = (ret == FI_EAVAIL) ?
+			efa_cq_readerr(&cq->cq_fid, &cq_err_entry, flags) :
+			-FI_EAVAIL;
+		if (OFI_UNLIKELY(ret < 0)) {
+			if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+				EFA_WARN(FI_LOG_CQ,
+					 "failed to read cq error: %ld\n", ret);
+			return;
+		}
+		ofi_cq_write_error(cq, &cq_err_entry);
+		return;
+	}
+
+	temp_cq_entry = (struct fi_cq_tagged_entry *)cq_entry;
+	for (i = 0; i < ret; i++) {
+		(flags & FI_SOURCE) ?
+			ofi_cq_write_src(cq, temp_cq_entry->op_context,
+					 temp_cq_entry->flags,
+					 temp_cq_entry->len,
+					 temp_cq_entry->buf,
+					 temp_cq_entry->data,
+					 temp_cq_entry->tag,
+					 src_addr[i]) :
+			ofi_cq_write(cq, temp_cq_entry->op_context,
+				     temp_cq_entry->flags,
+				     temp_cq_entry->len,
+				     temp_cq_entry->buf,
+				     temp_cq_entry->data,
+				     temp_cq_entry->tag);
+
+		temp_cq_entry = (struct fi_cq_tagged_entry *)
+				((uint8_t *)temp_cq_entry + efa_cq->entry_size);
+	}
+}
+
+void efa_ep_rdm_progress(struct util_ep *ep)
```

Comment:
This and the dgrm progress function are doing the same thing, can we combine them into one and register with ofi_endpoint_init()?